group_identifier_assigner.py

```diff
@@ -0,0 +1,262 @@
+    print(_usage, file=sys.stderr)
+    sys.exit()
+
+
+class Options:
+    endpoint = ENDPOINT
+    user = "co_8.william_test"
+    osg_co_id = OSG_CO_ID
+    outfile = None
+    authstr = None
+    min_timeout = MINTIMEOUT
+    max_timeout = MAXTIMEOUT
+    project_gid_startval = PROJECT_GIDS_START
+
+
+options = Options()
+
+
+def getpw(user, passfd, passfile):
+    if ":" in user:
+        user, pw = user.split(":", 1)
+    elif passfd is not None:
+        pw = os.fdopen(passfd).readline().rstrip("\n")
+    elif passfile is not None:
+        pw = open(passfile).readline().rstrip("\n")
+    elif "PASS" in os.environ:
+        pw = os.environ["PASS"]
+    else:
+        usage("PASS required")
+    return user, pw
+
+
+def mkauthstr(user, passwd):
+    from base64 import encodebytes
+
+    raw_authstr = "%s:%s" % (user, passwd)
+    return encodebytes(raw_authstr.encode()).decode().replace("\n", "")
+
+
+def mkrequest(method, target, data, **kw):
+    url = os.path.join(options.endpoint, target)
+    if kw:
+        url += "?" + "&".join("{}={}".format(k, v) for k, v in kw.items())
+    req = urllib.request.Request(url, json.dumps(data).encode("utf-8"))
+    req.add_header("Authorization", "Basic %s" % options.authstr)
+    req.add_header("Content-Type", "application/json")
+    req.get_method = lambda: method
+    return req
+
+
+def call_api(target, **kw):
+    return call_api2(GET, target, **kw)
+
+
+def call_api2(method, target, **kw):
+    return call_api3(method, target, data=None, **kw)
+
+
+def call_api3(method, target, data, **kw):
+    req = mkrequest(method, target, data, **kw)
+    trying = True
+    currentTimeout = options.min_timeout
+    while trying:
+        try:
+            resp = urllib.request.urlopen(req, timeout=currentTimeout)
+            payload = resp.read()
+            trying = False
+        except urllib.error.URLError as exception:
+            if currentTimeout < options.max_timeout:
+                currentTimeout *= TIMEOUTMULTIPLE
+            else:
+                sys.exit(
+                    f"Exception raised after maximum timeout {options.max_timeout} seconds reached. "
+                    + f"Exception reason: {exception.reason}.\n Request: {req.full_url}"
+                )
+
+    return json.loads(payload) if payload else None
+
+
+def get_osg_co_groups():
+    return call_api("co_groups.json", coid=options.osg_co_id)
+
+
+# primary api calls
+
+
+def get_co_group_identifiers(gid):
+    return call_api("identifiers.json", cogroupid=gid)
+
+
+def get_co_group_members(gid):
+    return call_api("co_group_members.json", cogroupid=gid)
+
+
+def get_co_person_identifiers(pid):
+    return call_api("identifiers.json", copersonid=pid)
+
+
+def get_datalist(data, listname):
+    return data[listname] if data else []
+
+
+def identifier_index(id_list, id_type):
+    found_list = list((id["Type"] == id_type for id in id_list))
+    if any(found_list):
+        return found_list.index(True)
+    return -1
```

Comment:
This works, but I think you can use a list comprehension directly for `found_list` (`[id["Type"] == id_type for id in id_list]`) rather than creating and consuming a generator, but even better, don't bother with any comparisons except for the `index` lookup using a `try`/`except` to retain the same return value behavior.
```suggestion
def identifier_index(id_list, id_type):
    id_type_list = [id["Type"] for id in id_list]
    try:
        return id_type_list.index(id_type)
    except ValueError:
        return -1
```

Reply:
I've incorporated this into [5dd0437](https://github.com/opensciencegrid/osg-comanage-project-usermap/pull/18/commits/5dd04376a4721ad1a9b7ff0511f4ed9f75284124)